create new branch get-keto-meal to get a random Meal that achieve the keto requirements

mohamedfawzy0 opened the pull request.

Commit:
create new branch get-keto-meal to get a random Meal that achieve the keto requirements

AimanYosofiUlt commented on src/main/kotlin/domain/model/Meal.kt:
```diff
@@ -51,3 +51,10 @@ data class Meal(
     }
 }
 
+fun Meal.isKetoMeal(ketoMealList: MutableSet<Meal>): Boolean {
```
Why do you add this extension function here?

AimanYosofiUlt commented on src/main/kotlin/domain/usecase/GetKetoMealUseCase.kt:
```diff
@@ -0,0 +1,19 @@
+        val ketoMeal = repo.getAllMeals().filter { meal ->
+            meal.isKetoMeal(ketoMealList)
+        }.randomOrNull() ?: throw kotlin.Exception("No more keto meals available.")
+        ketoMealList.add(ketoMeal)
```
What about adding a new function that stores the keto meal inside the `ketoMealList`, something like:
```
 val ketoMeal = repo.getAllMeals()
      .filter(::isKetoMeal)
      .randomOrNull() 
      .store()
      ?: throw kotlin.Exception("No more keto meals available.")
```

```
fun Meal.store(): Meal{
      ketoMealList.add(this)
      return this
}
```

AimanYosofiUlt reviewed (changes requested):
Need a little bit of refactoing and you are in the track to finish the task, Good job bro

Commit:
enhance in getKetoMeal function to work as functional approach

Commit:
add also function instead of store function

Ala-Sh reviewed (approved):
Well done!
I have nothing to suggest

AimanYosofiUlt commented on src/main/kotlin/domain/usecase/GetKetoMealUseCase.kt:
```diff
@@ -0,0 +1,25 @@
+    fun isKetoMeal(meal: Meal): Boolean {
+        val carbs = meal.nutrition?.carbohydrates ?: return false
+        val protein = meal.nutrition.protein ?: return false
+        return carbs <= 15.0 && protein >= 10.0
```
Please store these values inside constant variables, this is called a magic number

mohamedfawzy0 replied:
ok i solved it plz review

AimanYosofiUlt reviewed (changes requested):
please remove the magic value, and that's it, I will approve the changes after that

Commit:
enhance in the 2 use cases :
GetEasyPreparedMealsUseCase to use random index instead of shuflled
GetKetoMealUseCase to not use magic number anymore

Ala-Sh commented on src/main/kotlin/domain/usecase/GetEasyPreparedMealsUseCase.kt:
```diff
@@ -1,15 +1,58 @@
+        return result.toList()
+    }
+
+    /*fun List<Meal>.takeRandom(count: Int): List<Meal> {
```
Aiman will tell u to delete the commented code 😁

mohamedfawzy0 replied:
solved now quickly

Commit:
remove commented code

Ala-Sh commented on src/main/kotlin/domain/usecase/GetEasyPreparedMealsUseCase.kt:
```diff
@@ -1,15 +1,58 @@
+    fun List<Meal>.takeRandom(count: Int): List<Meal> {
+        if (isEmpty() || count <= 0) return emptyList()
+
+        val result = mutableSetOf<Meal>()
```
using a set impressed me
I think it's a genius idea🤩

mohamedfawzy0 replied:
learning from you 😁🤩

Commit:
Merge remote-tracking branch 'origin/development' into feature/get-keto-meal

Commit:
solve issue of repeated packages
